### PR TITLE
Migrate to PureScript 0.15

### LINKED ISSUE #79: Switch to esbuild
I noticed #62 , and you haven't considered esbuild seemingly.

https://esbuild.github.io/

This is the best bundler I know of.

### LINKED ISSUE #1430: Update `affjax` v12 to v13
`affjax` v12 is broken when bundled with webpack. It uses `module.require` to detect the environment(node or web), but webpack does not defined `module.require` resulting in it generating node bundles using `XMLHttpRequest` instead of `xhr2` node module, and requests failing with `Error: (ClientHttpError There was a problem making the request: XMLHttpRequest is not defined)`.

Looks like v13 allows environment detection to be handled by user by giving `affjax-web` and `affjax-node` modules. It would be nice to update `affjax` version.

### PR DESCRIPTION
**Pull request by errfrom**
Closes #79  
Closes #1430

1. Update `purs` to `v0.15.8`
2. Update `spago` to `v0.21.0`
3. Upgrade package-set
4. Remove deprecated `purescript-math` dependency
5. Remove every occurrence of using deprecated `MonadZero` type class
6. Upgrade `medea-ps` to PureScript 0.15 https://github.com/mlabs-haskell/medea-ps/pull/1
7. Drop `sequences` dependency https://github.com/mlabs-haskell/purescript-aeson/pull/18
8. Upgrade `purescript-bignumber` to PureScript 0.15 https://github.com/mlabs-haskell/purescript-bignumber/pull/3
9. Upgrade `purescript-aeson` to PureScript 0.15 https://github.com/mlabs-haskell/purescript-aeson/pull/17
10. Upgrade `noble-secp256k1` to PureScript 0.15 https://github.com/mlabs-haskell/purescript-noble-secp256k1/pull/3
11. Upgrade `toppokki` to PureScript 0.15 https://github.com/mlabs-haskell/purescript-toppokki/pull/1
12. Convert `cjs` exports to `esm` using [lebab](https://github.com/lebab/lebab), replace `export var` with `export const`
13. Fix `EncodeAeson` and `DecodeAeson` instances for `Ctl.Internal.Plutus.Types.AssocMap`
14. Remove `ToMetadata` instance for `(NonEmpty f a)` **TODO: Can it be rewritten instead?**
15. Use `AffjaxDriver` depending on the environment (node or browser)
16. Replace `SProxy` with kind-generic `Proxy`
17. Fix `ToData` / `FromData` instances for `Ctl.Internal.Types.Interval.Extended` and `Test.Data.Tree`
18. Use `Test.QuickCheck.Gen.frequency` with `NonEmptyArray`
19. Use `exists` from `Node.FS.Sync` (`Node.FS.Aff.exists` has been removed)
20. **Milestone reached: CTL compiles without errors**
21. Use `esm` imports
22. **FIXME**: Do not re-export `Undefined` from `Contract.Prelude` (remove `undefined` dependency from `spago.dhall`)
23. Fix CIP-30 Plutip tests, use `global` object in Node.js environment
24. **Milestone reached: All unit and Plutip tests pass**
25. **Add support for bundling using `esbuild`**
26. Create `node_modules` symlinks to `NODE_PATH` (see [NODE_PATH is not part of resolving import specifiers](https://nodejs.org/api/esm.html#esm_no_node_path))
27. **E2E tests pass** (`bundlePursProject` now uses `esbuild`)
28. Fix bundling using `webpack`
29. **TODO: fix and re-enable `@mlabs-haskell/csl-gc-wrapper`**

### Pre-review checklist

- [x] All code has been formatted using our config (`make format`)
- [x] Any new API features or modification of existing behavior is covered as defined in our [test plan](https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/doc/test-plan.md)
- [ ] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,7 +15,14 @@ result-*
 .node
 .node-cfg
 .projectile
-/dist/
+/dist/*
+!/dist/esbuild/
+/dist/esbuild/*
+!/dist/esbuild/index.html
```
Comment (klntsky) [inline .gitignore:21]: Please do not keep any files in dist. The name suggests it should be for build artifacts. It's OK to copy some data from e.g. `static` to `dist`.

File: dist/esbuild/index.html [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+<html lang="en" data-framework="purescript">
+<head>
+  <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
+</head>
+
+<body>
+  <script type="module">
+    import("./index.js").then(m => m.main());
+  </script>
+  Examples source code is located in <i>examples/</i> directory.<br>
```
Comment (klntsky) [inline dist/esbuild/index.html:11]: Let's put a link to this dir on github here

File: doc/getting-started.md [hunk 3]
```diff
@@ -38,7 +38,9 @@ Then, add `https://public-plutonomicon.cachix.org` and `public-plutonomicon.cach
 
 The easiest way to create a new CTL project is to use our `ctl-scaffold` flake template. This lives in the CTL repo -- you can have a look [here](../templates/ctl-scaffold). It contains a simple, yet complete, flakes-based scaffolding project with example `outputs` for a CTL project, including its runtime.
 
+<!-- markdown-link-check-disable -->
```
Comment (klntsky) [inline doc/getting-started.md:41]: I asked why didn't we keep it public

**Comment by klntsky** [comment]
> Hey folks! When can we expect this to be merged?

@bladyjoker It's getting delayed because I want to merge #1532 frist

**Comment by bladyjoker** [comment]
@klntsky friendly ping, a lot is relying on this PR ending up merged. Are there any blockers here?

**Comment by bladyjoker** [comment]
> @bladyjoker there are two blockers:
> 
>     1. releasing the version that is the current `develop`
> 
>     2. updating the changelog here, running full set of checks

Thanks @klntsky ! Can we prioritize this? LB is using this branch expecting it's about to get merged. LB users then use the same branch which then propagates the problem more :)

I just don't want for this PR to go stale, risking lots of effort going to waste. I think this is a big move for CTL and we should wrap it up? Are there any resources you're missing?

**Comment by klntsky** [comment]
> Are there any resources you're missing?

@bladyjoker time - was busy with catalyst proposal writing

**Comment by bladyjoker** [comment]
The last remaining issue we're tracking down is https://hercules-ci.com/accounts/github/Plutonomicon/derivations/%2Fnix%2Fstore%2Ffhn3q2yjdl6jq5m0lqcfpj4ld5lyrzlz-ctl-plutip-test.drv/log?via-job=d200f417-2c25-447d-bbff-2e0de43828f3

Which is the build result of https://github.com/Plutonomicon/cardano-transaction-lib/pull/1573

**Comment by klarkc** [comment]
@bladyjoker I've seen this same error in #1566, after debugging plutip cardano-node logs, I found an error which was being caused by a wrong cardano-api version in plutip, that's why I created https://github.com/mlabs-haskell/plutip/pull/184

**Comment by brainrake** [comment]
This is an heroic effort and I appreciate it.

**Comment by bladyjoker** [comment]
> @bladyjoker I've seen this same error in #1566, after debugging plutip cardano-node, I found an error which was being caused by a wrong cardano-api version in plutip, that's why I created [mlabs-haskell/plutip#184](https://github.com/mlabs-haskell/plutip/pull/184)

@klarkc we managed to get everything working centered around cardano-node v8.1.1. 

Moving to v8.5 will have to be a battle for some other day. Can you summarize the work you did across CTL and Plutip and whether your efforts for 8.5 migration resulted in passing tests or were there some lingering issues?

**Comment by klarkc** [comment]
@bladyjoker the last time I checked, we're being blocked by [plutip -> cardano-wallet](https://github.com/mlabs-haskell/plutip/pull/184#issuecomment-1840748353). We still need a valid build plan for plutip with cardano-wallet.

**Comment by klntsky** [comment]
Building searchable documentation fails. Not merging yet.

**Review by bladyjoker — APPROVED** [review]
Just wanted to click this button after all this work :)

**Comment by klarkc** [comment]
BTW, congratulations to everyone on achieving it. It will really help us here.
